File: 368__Largest_Divisible_Subset.py
class Solution(object):
    def largestDivisibleSubset(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        nums.sort()
        dp = [[nums[i]] for i in range(len(nums))]
        for i in range(1, len(nums)):
            for j in range(i-1, -1, -1):
                if nums[i] % nums[j] == 0:
                    if len(dp[j]) + 1 > len(dp[i]):
                        dp[i] = list(dp[j])
                        dp[i].append(nums[i])
        maxcount = 0
        result = []
        for lst in dp:
            if len(lst) > maxcount:
                result = lst
                maxcount = len(lst)
        return result

File: Java/368__Largest_Divisible_Subset.java
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        if (nums.length == 0) return new ArrayList<Integer>();
        Arrays.sort(nums);
        int[] sizes = new int[nums.length];
        int[] prevs = new int[nums.length];
        int maxidx = 0;
        for (int i = 0; i < nums.length; i++) {
            sizes[i] = 1;
            prevs[i] = -1;
            for (int j = 0; j < i; j++) {
                if (nums[i] % nums[j] == 0 && sizes[i] < sizes[j] + 1) {
                    sizes[i] = sizes[j] + 1;
                    prevs[i] = j;
                }
            }
            if (sizes[i] > sizes[maxidx]) maxidx = i;
        }

        List<Integer> list = new ArrayList<>();
        for (int i = maxidx; i != -1; i = prevs[i]) {
            list.add(nums[i]);
        }
        return list;
    }
}
